controllers/actions.summerwind.net/horizontal_runner_autoscaler_batch_scale.go

```diff
@@ -179,6 +179,14 @@ func (s *batchScaler) batchScale(ctx context.Context, batch batchScaleOperation)
 				}
 			}
 
+			// Update the latest CapacityReservation time to now to trigger reconcile
+			// Without this, we might not scale up after an ephemeral runner has been deleted
+			// until a new scale up, all runners finish, or after DefaultRunnerPodRecreationDelayAfterWebhookScale
+			if len(reservations) > 0 {
+				reservations[len(reservations)-1].EffectiveTime = metav1.Time{Time: now}
+				reservations[len(reservations)-1].ExpirationTime = metav1.Time{Time: now.Add(scale.trigger.Duration.Duration)}
+			}
```

Comment:
@Smirl Great! This look hacky at glance but would fix the original issue.

If we are to make it more complete, I guess we might be able to move this logic to L178, so that every CapacityRervation after the removed one has updated EffectiveTime and ExpirationTime.

It might be a bit computationally costly because you now need to modify almost all the remaining capacity reservations slice items on each "completed" or "cancelled"  webhook event.

However, doing so would additionally fix another long-standing issue that a CapacityReservation can expire before the corresponding runner is deployed onto the cluster, when the runner creation was delayed due to reaching maxReplicas.

WDYT?

Reply:
[f6f14c8](https://github.com/actions/actions-runner-controller/pull/2258/commits/f6f14c824902c6ef0f07af2ae681da1765489aaa) is my attempt to address the above idea.

Reply:
Thanks for pushing the change.

From what I can see this now means that we will have a list of objects that always have the same `effectiveTime`, `expirationTime`, and _mostly/always_ `replicas: 1`. Could now be just `spec.effectiveTime`, `spec.expireationTime`, `spec.queuedReplicas` (or something like this).

I also noticed that the size of the object with ~10000 things in the queue exceeds the etcd 1Mb limit. Also, every change to this object (`spec` or `status`) will create a copy in etcd, meaning that large objects changing often could fill etcd space until Compaction kicks in.

Finally, happy to keep this as you have it for now as I imagine this is all going in the bin when the new `AutoscalingRunnerSet` becomes standard?

Reply:
@Smirl I greatly appreciate your detailed feedback and comment!

> From what I can see this now means that we will have a list of objects that always have the same effectiveTime, expirationTime, and mostly/always replicas: 1.

Correct!

>  Could now be just spec.effectiveTime, spec.expireationTime, spec.queuedReplicas (or something like this).

That's a good call, but I'm still uncertain if that's the right path forward according to your subsequent comment:

> I also noticed that the size of the object with ~10000 things in the queue exceeds the etcd 1Mb limit. Also, every change to this object (spec or status) will create a copy in etcd, meaning that large objects changing often could fill etcd space until Compaction kicks in.

Good catch! I'm aware of that, and that's the hidden defect of the current approach.

All in all, perhaps we can just merge all the capacity reservations into one, and "reset" the sole, aggregated capacity reservation's effective and expiration times on scale down(i.e. on receiving workflow job event of status=completed).

That would reduce the number of things down to 1 from 10000, with all the behaviors we want (reset the expiration and effective times on scale down when max replicas is configured when there are reserved but not created runner remaining in the capacity reservation).

WDYT?

Reply:
> Finally, happy to keep this as you have it for now as I imagine this is all going in the bin when the new AutoscalingRunnerSet becomes standard?

I have mixed feelings. Yeah we might just defer the resolution to the autoscaling runner set and say HRA won't scale well for more than hundreds of runners. Or, shall I merge this anyway, cut ARC 0.27.1, and later do the capacity reservation merge thing in another PR? 🤔

Reply:
From reading the [preview docs](https://github.com/actions/actions-runner-controller/tree/master/docs/preview/actions-runner-controller-2) it seems like this approach is better? Will it have the same issues?

I like the merge now, consolidate later approach. I know that we would be unblocked with a 0.27.1 release as we currently only have a few repos using this controller (but will have 1000s by the end of the year 😱 ).

Regarding merging the capacity reservations into 1, I _think_ it would be safer to do this in another PR.